src/lib/SaferNames/SourceRename.hs

```diff
@@ -0,0 +1,368 @@
+  sourceRenameB NothingB = returnG NothingB
+
+sourceRenameUBinderNest :: Renamer m => (forall o'. Name s o' -> SourceNameDef o')
+                        -> Nest (UBinder s) i i' -> RenamerNameGenT m (Nest (UBinder s) o) o
+sourceRenameUBinderNest _ Empty = returnG Empty
+sourceRenameUBinderNest f (Nest b bs) =
+  sourceRenameUBinder f b `bindG` \b' ->
+  sourceRenameUBinderNest f bs `bindG` \bs' ->
+  returnG $ Nest b' bs'
+
+sourceRenameUBinder :: Renamer m => (forall o'. Name s o' -> SourceNameDef o')
+                    -> UBinder s i i' -> RenamerNameGenT m (UBinder s o) o
+sourceRenameUBinder asSourceNameDef ubinder = case ubinder of
+  UBindSource b -> RenamerNameGenT do
+    SourceMap sourceMap <- askSourceMap
+    mayShadow <- askMayShadow
+    unless (mayShadow || not (M.member b sourceMap)) $
+      throw RepeatedVarErr $ pprint b
+    withFreshM b \freshName -> do
+      (Distinct _) <- getScope
+      let frag = (singletonScope freshName)
+      let sourceMap' = SourceMap (M.singleton b (asSourceNameDef $ nameBinderName freshName))
+      return $ RenamerContent frag sourceMap' $ UBind freshName
+  UBind _ -> error "Shouldn't be source-renaming internal names"
+  UIgnore -> returnG UIgnore
+
+instance SourceRenamableE UDataDef where
+  sourceRenameE (UDataDef (tyConName, paramBs) dataCons) = do
+    (RenamerContent frag sourceMap paramBs') <- runRenamerNameGenT $ sourceRenameB paramBs
+    extendScope frag $ extendSourceMap sourceMap $ do
+      dataCons' <- forM dataCons \(dataConName, argBs) -> do
+        argBs' <- sourceRenameE argBs
+        return (dataConName, argBs')
+      return $ UDataDef (tyConName, paramBs') dataCons'
+
+instance SourceRenamableE UDataDefTrail where
+  sourceRenameE (UDataDefTrail args) = withSourceRenameB args \args' ->
+    return $ UDataDefTrail args'
+
+instance (SourceRenamableE e1, SourceRenamableE e2) => SourceRenamableE (PairE e1 e2) where
+  sourceRenameE (PairE x y) = PairE <$> sourceRenameE x <*> sourceRenameE y
+
+instance SourceRenamableE e => SourceRenamableE (ListE e) where
+  sourceRenameE (ListE xs) = ListE <$> mapM sourceRenameE xs
+
+instance SourceRenamableE UMethodDef where
+  sourceRenameE (UMethodDef v expr) =
+    UMethodDef <$> sourceRenameE v <*> sourceRenameE expr
+
+instance SourceRenamableB b => SourceRenamableB (Nest b) where
+  sourceRenameB (Nest b bs) =
+    sourceRenameB b `bindG` \b' ->
+    sourceRenameB bs `bindG` \bs' ->
+    returnG $ Nest b' bs'
+  sourceRenameB Empty = returnG Empty
+
+-- -- === renaming patterns ===
+
+-- We want to ensure that pattern siblings don't shadow each other, so we carry
+-- the set of in-scope siblings' names along with the normal renaming env.
+
+type SiblingSet = S.Set SourceName
+
+-- TODO It seems this monad actually has no way for a downstream action
+-- to check for membership in the sibling set produced by an upstream action?
+-- Do we already have an ask/extend type of monad for this sort of thing?
```

Comment:
Yes, good observation. We should make the NameGen instance take a constraint that the underlying monad gives you read-access to the sibling set, like we do here: https://github.com/google-research/dex-lang/blob/dev/src/lib/SourceRename.hs#L313